aten/src/ATen/autocast_mode.h

```diff
@@ -72,13 +81,88 @@ inline at::ScalarType promote_type(at::ScalarType current, Arg0 arg0, Args... ar
   return promote_type(new_current, args...);
 }
 
+// Overload prioritize with extra DeviceType parameter
+inline at::ScalarType prioritize(
+    DeviceType device_type,
+    at::ScalarType current,
+    const Tensor& nextArg) {
+  if (current == at::kDouble) {
+    AT_ERROR("promote type is double in at::autocast::prioritize");
+    return current;
+  }
+  if (is_autocast_eligible(device_type, nextArg)) {
+    auto next = nextArg.scalar_type();
+    if (next == at::kDouble) {
+      return current; // ignores double tensors
+    } else if (current == at::kFloat || next == at::kFloat) {
+      return at::kFloat; // prioritizes float over half
+    } else if (current == at::kHalf && next == at::kHalf) {
+      return at::kHalf;
+    } else if (current == at::kBFloat16 && next == at::kBFloat16) {
+      return at::kBFloat16;
+    } else {
+      AT_ERROR("Unexpected floating ScalarType in at::autocast::prioritize");
+      return current;
+    }
+  } else {
+    return current;
+  }
+}
```

Comment:
A lot more unnecessary duplication here again

Reply:
Same as the solution of `cached_cast` add a DeviceType argument with default device parameter of CUDA for the function of `prioritize`, then we remove all the duplication. As for the function of `promote_type`, the changes is not BC. But I think I have audited all the invoked positions.